File: logistic_regression.py
import numpy as np
from exceptions import RuntimeError

import optimization as opt
from auxillary import sigmoid

class LogisticRegression:
    """A basic implementation of logistic regression."""
    def __init__(self):
        self.theta_trained = None
        self.num_iters = None
        self.vtheta_err = None
        self.vcost = None

    def predict(self, X, theta=None, threshold=0.5):
        """Predict output based on observations X and linear weights theta."""
        theta = self.theta_trained if theta is None else theta
        if theta is None:
            raise RuntimeError('This model has not been trained.')

        return np.array(sigmoid(np.dot(X,theta)) >= threshold, dtype=int)
    
    def cost(self, X, theta, y):
        """Calculate the cost/objective function given observations X, linear weights theta, and target y."""
        m = X.shape[0]
        h = sigmoid(np.dot(X,theta))
        return -(np.dot(y,np.log(h)) + np.dot((1-y),np.log(1-h)))/m

    def cost_grad(self, X, theta, y):
        """Calculate the gradient of the cost/objective function at theta."""
        m = X.shape[0]
        h = sigmoid(np.dot(X,theta))
        return np.dot(X.T, h - y)/m

    def train(self, X, y, tol=0.0001, max_iters=500, verbose=False):
        """Train on data X with target y."""
        def grad_func(theta): return self.cost_grad(X,theta,y)
        if not verbose:
            self.theta_trained,self.num_iters = opt.gradient_descent(grad_func,
                                                                     0.1,
                                                                     np.zeros(X.shape[1]),
                                                                     max_iters,
                                                                     tol)
        else:
            def cost_func(theta): return self.cost(X,theta,y)
            self.theta_trained,self.num_iters,self.vtheta_err,self.vcost = opt.gradient_descent_verbose(
                grad_func,
                0.1,
                np.zeros(X.shape[1]),
                max_iters,
                tol,
                cost_func)
            
        return self.theta_trained

File: LogisticRegression.jl
module LogisticRegression

export LogitRegression
export train!, predict

using Optimization
using Regularization

abstract Model

"Simple logistic regression model"
type LogitRegression <: Model
    theta::Array{Float64,1}
    thresh::Float64
    lmbda::Float64
    alpha::Float64
end

"Sigmoid function"
sigmoid(z) = 1.0./(1.0 + exp(-z))


# Cost Functions
# --------------
"""
    cost_logit(theta, lmbda, X, y)

Compute the cost function associated with logistic regression.
"""
function cost_logit(theta, lmbda, X, y)
    m = size(X)[1]
    h = sigmoid(X*theta)
    (-y'*log(h) - (1-y)'*log(1-h))/m + penalty(theta,lmbda)
end



"""
    cost_grad_logit(theta, lmbda, X, y)

Compute the gradient of the cost function for lostic regression.
"""
function cost_grad_logit(theta, lmbda, X, y)
    m = size(X)[1]
    err = sigmoid(X*theta) - y
    (X'*err)/m + penalty_grad(theta,lmbda)
end

cost(model::LogitRegression) = cost_logit
cost_grad(model::LogitRegression) = cost_grad_logit

# Training and predicting functions
# ---------------------------------
"""
    train!(model, X, y)

Train model using feature matrix X and labels y.

The params argument provides the parameter input to the
optimization routine.  The objective function is added
to these parameters, but it is only called if a callback
function is also provided.
"""
function train!(model::LogitRegression, X, y; params...)

    cost_model = cost(model)
    cost_grad_model = cost_grad(model)
    
    objective(theta) = cost_model(theta, model.lmbda, X, y)
    grad(theta) = cost_grad_model(theta, model.lmbda, X, y)

    params = Dict{Any,Any}(params)
    if !(:objective in keys(params))
        params[:objective] = objective
    end
    
    model.theta = gradient_descent(grad, model.alpha, model.theta; params...)
end

"""
    predict(model, X)

Make a prediction on feature matrix X given model.
"""
function predict(model::LogitRegression, X)
    h = sigmoid(X*model.theta)
    [hi >= model.thresh ? 1 : 0 for hi in h]
end

end
